### DISCUSSION TITLE
How to detect broken WebSocket connection?

### DISCUSSION THREAD

**Discussion by a-riley-ulab** [curl/curl#9886] [Q&A] [answered]
It seems like `curl_ws_recv` has no way to report if the underlying socket is closed or broken. Internally, it uses `curl_easy_recv`, and [in the documentation](https://curl.se/libcurl/c/curl_easy_recv.html) for that function, it indicates that receiving zero bytes without a CURLE_AGAIN means the connection is broken or closed. However, `curl_ws_recv` will report zero bytes in two situations:
1. There was received data, but no message (e.g. a ping from a server with no message).
2. The connection is dead and `curl_easy_recv` indicated that with zero bytes read.

So in this case, it is impossible to disambiguate at the call site which of these happened. Is there another way to tell if the connection is dead? I'd like to avoid reaching into raw system sockets if possible. I have an internal fork where `curl_ws_recv` will return `CURL_RECV_ERROR` in situation 2, but I'm not sure if this is how it should be done.

**Comment by MonkeybreadSoftware** [comment]
I just played today with websocket functions (what a lovely way to spend a Sunday).
I run into a problem, that curl_ws_recv blocks potential forever and I thought my app had a bug.
So I see a problem, that curl_ws_recv does an endless while loop. it should do have a timeout and give up after some time.

We also came to the idea to changing the underlaying socket to non-blocking would fix that. But now we can't really know whether the socket is dead or closed.
I would welcome, if curl_ws_recv could return an error, if the socket is broken or closed.

**Reply by bagder** [reply]
>  I would welcome, if curl_ws_recv could return an error, if the socket is broken or closed.

makes sense. I'll work on creating a test case for this and make sure it works.

**Comment by bagder** [comment, marked as answer]
I made #10084, which makes `curl_ws_recv()` return `CURLE_GOT_NOTHING` on connection close.

You think this works?

**Reply by a-riley-ulab** [reply]
Looks good to me from a quick test. I should have mentioned that I was also using the non-blocking mode. Thanks!
